qiskit/synthesis/boolean/boolean_expression_synth.py

```diff
@@ -0,0 +1,126 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2025.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""Circuit synthesizers and related classes for boolean expressions"""
+
+import itertools
+from qiskit.circuit import QuantumCircuit
+from qiskit.circuit.library import ZGate, XGate
+
+
+class EsopGenerator:
+    """Generates an ESOP (Exlusive-sum-of-products) representation
+    for a boolean function given by its truth table"""
+
+    def __init__(self, truth_table):
+        self.truth_table = truth_table
+        self.cofactor_table = {}
+        self.esop = self.generate_esop(assignment=())
+
+    def clause_diff(self, clause1, clause2):
+        """The indices of variables where the clauses differ"""
+        return [i for i in range(len(clause1)) if clause1[i] != clause2[i]]
+
+    def combine_and_optimize(self, clauses_1, clauses_2):
+        """Combining clauses of distance 1 until no more combinations can be performed"""
+        something_changed = True
+        combined_clauses = clauses_1 + clauses_2
+        while something_changed:
+            something_changed = False
+            for clause_1, clause_2 in itertools.combinations(combined_clauses, 2):
+                diff = self.clause_diff(clause_1, clause_2)
+                if len(diff) == 1:
+                    new_clause = clause_1[: diff[0]] + "-" + clause_1[diff[0] + 1 :]
+                    if new_clause not in combined_clauses:
+                        combined_clauses.remove(clause_1)
+                        combined_clauses.remove(clause_2)
+                        combined_clauses.append(new_clause)
+                        something_changed = True
+                        break
+        return combined_clauses
+
+    def generate_esop(self, assignment):
+        """Recursively generates an ESOP for a partially determined boolean expression
+        'assignment' is a partial assignment to the expression's variables
+        """
+        # this base case is non-optimized;
+        # we may be able to terminate even before having a full assignment
+        if len(assignment) == self.truth_table.num_bits:
+            if self.truth_table[assignment]:
+                return ["-" * self.truth_table.num_bits]  # True; a don't care clause
```

Comment:
I don't understand this line. Let's say the assignment is 101. Why does this return "---" and not "101"?

Reply:
This is the base case for a function that was already fully specified, assignment-wise, so now it's a constant. If it's the constant `T`, then the assignment doesn't matter. Later, when returning from the recursion base, we add specifications into the clause. See [here](https://people.csail.mit.edu/costis/papers/esop.pdf) for more details (what we do here is basically what the paper refers to as the Shannon expansion).

Reply:
Thanks, I see, all these `-` values will be later replaced by either `0` or `1`, so we could have just as well put any other symbol (e.g., `*`) there instead of `-`. So my only (minor) concern is that if in the future the algorithm for generating ESOPs changes, we don't mix up between `-` symbols that mean "the value will later be overwritten" and between `-` symbols that mean "the corresponding variable has already been expanded and the value can be both `0` and `1`". Maybe you can put indeed `"*"` there to avoid such potential future, or add a comment in code that explains this point.

Reply:
On the contrary; I think that using `*` might cause future problems, that is why I used `-`.

In the Shannon expansion, the idea is to recursively generate ESOP's for partially-assigned functions ("subfunctions"). For the subfunction which is always `T`, the correct ESOP is indeed one clause of `-` for all variables. So it's not here as a placeholder but as the actual base case.

Note, if you look at the note I refenced above, you can see that one might use other expansions (e.g. positive and negative Davio) which use the ESOP for a subfunction but does not fill it completely. Hence, if we extend the current code to handle Davio as well, the correct way to go is to have `-` clauses.

Reply:
I agree now -- thanks for clarifying this!